pytential/qbx/__init__.py

```diff
@@ -87,111 +90,135 @@ class QBXLayerPotentialSource(LayerPotentialSourceBase):
+            :meth:`boxtree.traversal.FMMTraversalBuilder.__call__`.
         :arg _use_target_specific_qbx: Whether to use target-specific
-            acceleration by default if possible. *None* means
-            "use if possible".
+            acceleration by default if possible. *None* means "use if possible".
+
         :arg cost_model: Either *None* or an object implementing the
              :class:`~pytential.qbx.cost.AbstractQBXCostModel` interface, used for
-             gathering modeled costs if provided (experimental)
+             gathering modeled costs if provided (experimental).
         """
 
         # {{{ argument processing
 
         if fine_order is None:
-            raise ValueError("fine_order must be provided.")
+            raise ValueError("'fine_order' must be provided.")
+        assert isinstance(fine_order, int)
 
         if qbx_order is None:
-            raise ValueError("qbx_order must be provided.")
-
-        if target_stick_out_factor is not _not_provided:
-            from warnings import warn
-            warn("target_stick_out_factor has been renamed to "
-                    "target_association_tolerance. "
-                    "Using target_stick_out_factor is deprecated "
-                    "and will stop working in 2018.",
-                    DeprecationWarning, stacklevel=2)
-
-            if target_association_tolerance is not _not_provided:
-                raise TypeError("May not pass both target_association_tolerance and "
-                        "target_stick_out_factor.")
-
-            target_association_tolerance = target_stick_out_factor
-
-        del target_stick_out_factor
+            raise ValueError("'qbx_order' must be provided.")
+        assert isinstance(qbx_order, int)
 
         if target_association_tolerance is _not_provided:
-            target_association_tolerance = float(
-                    np.finfo(density_discr.real_dtype).eps) * 1e3
+            target_association_tolerance = (
+                1.0e+3 * float(np.finfo(density_discr.real_dtype).eps))
+        assert isinstance(target_association_tolerance, float)
 
         if fmm_order is not None and fmm_level_to_order is not None:
-            raise TypeError("may not specify both fmm_order and fmm_level_to_order")
+            raise TypeError(
+                "may not specify both 'fmm_order' and 'fmm_level_to_order'.")
 
         if _box_extent_norm is None:
             _box_extent_norm = "l2"
+        assert isinstance(_box_extent_norm, str)
 
         if _from_sep_smaller_crit is None:
             # This seems to win no matter what the box extent norm is
             # https://gitlab.tiker.net/papers/2017-qbx-fmm-3d/issues/10
             _from_sep_smaller_crit = "precise_linf"
+        assert isinstance(_from_sep_smaller_crit, str)
 
         if fmm_level_to_order is None:
-            if fmm_order is False:
+            if fmm_order is False or fmm_order is None:
```

Comment:
Not sure what was meant to happen when both `fmm_level_to_order` and `fmm_order` are `None`?

Reply:
Good question. I'd say error, we'll see how badly that blows up.

Reply:
It did that before (because I forgot to check for `None`) and there are just 8 errors:
https://github.com/inducer/pytential/runs/7228738492?check_suite_focus=true

That's not so bad, so I'll just fix those up and raise an error.